Mathlib/CategoryTheory/ChosenFiniteProducts.lean

```diff
@@ -195,6 +195,263 @@ instance (priority := 100) : Limits.HasFiniteProducts C :=
+    ext z
+    apply hom_ext <;>
+    simp only [Functor.comp_obj, whiskeringRight_obj_obj, curriedTensor_obj_obj,
+      whiskeringLeft_obj_obj, Functor.comp_map, whiskeringRight_obj_map, NatTrans.comp_app,
+      whiskerRight_app, curriedTensor_map_app, prodComparisonNatTrans_app, Category.assoc,
+      prodComparison_fst, whiskeringLeft_obj_map, whiskerLeft_app, whiskerRight_fst,
+      prodComparison_fst_assoc, prodComparison_snd, whiskerRight_snd, ← F.map_comp]
+
+
+instance (A : C) [∀ B, IsIso (prodComparison F A B)] : IsIso (prodComparisonNatTrans F A) := by
+  letI : ∀ X, IsIso ((prodComparisonNatTrans F A).app X) := by assumption
+  apply NatIso.isIso_of_isIso_app
+
+instance [h : ∀ A B, IsIso (prodComparison F A B)] : IsIso (prodComparisonBifunctorNatTrans F) := by
+  letI : ∀ X, IsIso ((prodComparisonBifunctorNatTrans F).app X) :=
+    fun _ ↦ (by dsimp; apply NatIso.isIso_of_isIso_app)
+  apply NatIso.isIso_of_isIso_app
+
+/-- The natural isomorphism `F(A ⊗ -) ≅ FA ⊗ F-`, provided each `prodComparison F A B` is an
+isomorphism (as `B` changes). -/
+@[simps!]
+noncomputable def prodComparisonNatIso (A : C)
+    [∀ B, IsIso (prodComparison F A B)] :
+    (curriedTensor C).obj A ⋙ F ≅ F ⋙ (curriedTensor D).obj (F.obj A) :=
+  asIso (prodComparisonNatTrans F A)
+
+@[simp]
+lemma prodComparisonNatIso_hom (A : C) [∀ B, IsIso (prodComparison F A B)] :
+    (prodComparisonNatIso F A).hom = prodComparisonNatTrans F A := rfl
+
+/-- The natural isomorphism of bifunctors `F(- ⊗ -) ≅ F- ⊗ F-`, provided each
+`prodComparison F A B` is an isomorphism. -/
+@[simps!]
+noncomputable def prodComparisonBifunctorsNatIso
+    [∀ A B, IsIso (prodComparison F A B)] :
+    curriedTensor C ⋙ (whiskeringRight _ _ _).obj F ≅
+      F ⋙ curriedTensor D ⋙ (whiskeringLeft _ _ _).obj F :=
+  asIso (prodComparisonBifunctorNatTrans F)
+
+@[simp]
+lemma prodComparisonBifunctorsNatIso_hom [∀ A B, IsIso (prodComparison F A B)] :
+    (prodComparisonBifunctorsNatIso F).hom = prodComparisonBifunctorNatTrans F := rfl
+
+theorem prodComparison_comp {E : Type u₂} [Category.{v₂} E] [ChosenFiniteProducts E] (G : D ⥤ E) :
+    prodComparison (F ⋙ G) A B =
+      G.map (prodComparison F A B) ≫ prodComparison G (F.obj A) (F.obj B) := by
+  unfold prodComparison
+  ext <;> simp <;> rw [← G.map_comp] <;> simp
+
+open Limits
+
+section PreservesLimitPairs
+
+variable (A B)
+variable [PreservesLimit (pair A B) F]
+
+/-- If `F` preserves the limit of the pair `(A, B)`, then the binary fan given by
+`(F.map fst A B, F.map (snd A B))` is a limit cone. -/
+def isLimitChosenFiniteProductsOfPreservesLimits :
+    IsLimit <| BinaryFan.mk (F.map (fst A B)) (F.map (snd A B)) :=
+  mapIsLimitOfPreservesOfIsLimit F (fst _ _) (snd _ _) <|
+    (product A B).isLimit.ofIsoLimit <| isoBinaryFanMk (product A B).cone
+
+/-- If `F` preserves the limit of the pair `(A, B)`, then `prodComparison F A B` is an isomorphism.
+-/
+def prodComparisonIso : F.obj (A ⊗ B) ≅ F.obj A ⊗ F.obj B :=
```

Comment:
I would that the standard form of the assumptions that a functor "preserves" chosen finite products should be using `PreservesLimit...`. Then, for future applications, it may be better to define `prodComparisonNatIso ` and `prodComparisonBifunctorNatIso` under `PreservesLimit` assumption, rather than `IsIso` assumptions, do not you think?

Reply:
(Otherwise, `prodComparisonIso` should not very much be a definition, but rather a step in the proof of `isIso_prodComparison_of_preservesLimit_pair`.)

Reply:
I wanted to stay as close as possible to the API for usual products that uses `IsIso` for everything, but I guess it’s indeed better to have `PreservesLimits` assumptions instead.